CONFIG.md
[![Website](https://img.shields.io/badge/-Website-lightblue.svg?longCache=true&style=for-the-badge&logo=data:image/svg%2bxml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz4KPHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI0MjAiCmhlaWdodD0iNDIwIiBzdHJva2U9IiMwMDAiIGZpbGw9Im5vbmUiPgo8cGF0aCBzdHJva2Utd2lkdGg9IjI2IgpkPSJNMjA5LDE1YTE5NSwxOTUgMCAxLDAgMiwweiIvPgo8cGF0aCBzdHJva2Utd2lkdGg9IjE4IgpkPSJtMjEwLDE1djM5MG0xOTUtMTk1SDE1TTU5LDkwYTI2MCwyNjAgMCAwLDAgMzAyLDAgbTAsMjQwIGEyNjAsMjYwIDAgMCwwLTMwMiwwTTE5NSwyMGEyNTAsMjUwIDAgMCwwIDAsMzgyIG0zMCwwIGEyNTAsMjUwIDAgMCwwIDAtMzgyIi8+Cjwvc3ZnPg==)](https://orbitalsync.com)
[![GitHub](https://img.shields.io/badge/-GitHub-lightgrey.svg?longCache=true&style=for-the-badge&logo=github)](https://github.com/mattwebbio/orbital-sync)
[![GitHub Stars](https://img.shields.io/github/stars/mattwebbio/orbital-sync?style=for-the-badge&logo=github&labelColor=lightgrey&color=lightgrey)](https://github.com/mattwebbio/orbital-sync)

[Installation](https://orbitalsync.com/#getting-started) | [Configuration](https://orbitalsync.com/CONFIG.html) | [Changelog](CHANGELOG.md)

<img src="https://user-images.githubusercontent.com/420820/187585158-331400c3-18f3-4673-857e-44efd73c6104.svg" width="200" alt="Logo" />

# Orbital Sync: Configuration

See [additional notes](#additional-notes) at the bottom for information such as how to use [Docker secrets](#docker-secrets).

<!-- START CONFIG DOCS -->

| Environment Variable | Required | Default | Example                                                         | Description                                                                                                                                                           |
| -------------------- | -------- | ------- | --------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `PI_HOLE_VERSION`    | No       | `auto`  | `auto`, `6`, or `5`                                             | The version of PiHole you are using                                                                                                                                   |
| `UPDATE_GRAVITY`     | No       | `true`  | `true`/`false`                                                  | Triggers a gravity update after a backup has been uploaded to a secondary Pi-hole. This updates adlists and restarts gravity.                                         |
| `VERBOSE`            | No       | `false` | `true`/`false`                                                  | Increases the verbosity of log output. Useful for debugging.                                                                                                          |
| `RUN_ONCE`           | No       | `false` | `true`/`false`                                                  | By default, Orbital Sync runs indefinitely until stopped. Setting this to `true` forces it to exit immediately after the first sync.                                  |
| `INTERVAL_MINUTES`   | No       | `60`    | Any non-zero positive integer, for example `5`, `30`, or `1440` | How long to wait between synchronizations. Defaults to sixty minutes. Remember that the DNS server on your secondary servers restarts every time a sync is performed. |

## Primary Host

The primary Pi-hole that data will be copied from.

| Environment Variable    | Required | Default | Example                                              | Description                                                                                                                       |
| ----------------------- | -------- | ------- | ---------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------- |
| `PRIMARY_HOST_BASE_URL` | Yes      | N/A     | `http://192.168.1.2` or `https://pihole.example.com` | The base URL of your Pi-hole, including the scheme (HTTP or HTTPS) and port but not including a following slash.                  |
| `PRIMARY_HOST_PASSWORD` | Yes      | N/A     | `mypassword`                                         | The password (v5) or app password (v6) used to login to PiHole.                                                                   |
| `PRIMARY_HOST_PATH`     | No       | N/A     | `/` or `/apps/pi-hole`                               | The path to be appended to your base URL. The default Pi-hole path is `/admin` (v5) or `/api` (v6), which is added automatically. |

## Secondary Hosts

Secondary Pi-holes that data will be copied to.

Replace `(#)` with a number, starting at 1, to add multiple. Each must be sequential, (i.e. `SECONDARY_HOSTS_1_BASE_URL`, `SECONDARY_HOSTS_2_BASE_URL`, `SECONDARY_HOSTS_3_BASE_URL`, and so on) and start at number 1. Any gaps (for example, 3 to 5 skipping 4) will result in configuration after the gap being skipped.

| Environment Variable           | Required | Default | Example                                               | Description                                                                                                                |
| ------------------------------ | -------- | ------- | ----------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------- |
| `SECONDARY_HOSTS_(#)_BASE_URL` | Yes      | N/A     | `http://192.168.1.3` or `https://pihole2.example.com` | The base URL of your secondary Pi-hole, including the scheme (HTTP or HTTPS) and port but not including a following slash. |
| `SECONDARY_HOSTS_(#)_PASSWORD` | Yes      | N/A     | `mypassword2`                                         | The password used to log in to the admin interface.                                                                        |
| `SECONDARY_HOSTS_(#)_PATH`     | No       | N/A     | `/` or `/apps/pi-hole`                                | The path to be appended to your secondary base URL. The default Pi-hole path is `/admin`, which is added automatically.    |

## Sync

What data to copy from the primary Pi-hole to the secondary Pi-holes.

### V5

Sync options for Pi-hole v5.x.

| Environment Variable          | Required | Default | Example        | Description                                                 |
| ----------------------------- | -------- | ------- | -------------- | ----------------------------------------------------------- |
| `SYNC_V5_WHITELIST`           | No       | `true`  | `true`/`false` | Copies the whitelist                                        |
| `SYNC_V5_REGEX_WHITELIST`     | No       | `true`  | `true`/`false` | Copies the regex whitelist                                  |
| `SYNC_V5_BLACKLIST`           | No       | `true`  | `true`/`false` | Copies the blacklist                                        |
| `SYNC_V5_REGEX_LIST`          | No       | `true`  | `true`/`false` | Copies the regex blacklist                                  |
| `SYNC_V5_AD_LIST`             | No       | `true`  | `true`/`false` | Copies adlists                                              |
| `SYNC_V5_CLIENT`              | No       | `true`  | `true`/`false` | Copies clients                                              |
| `SYNC_V5_GROUP`               | No       | `true`  | `true`/`false` | Copies groups                                               |
| `SYNC_V5_AUDIT_LOG`           | No       | `false` | `true`/`false` | Copies the audit log                                        |
| `SYNC_V5_STATIC_DHCP_LEASES`  | No       | `false` | `true`/`false` | Copies static DHCP leases                                   |
| `SYNC_V5_LOCAL_DNS_RECORDS`   | No       | `true`  | `true`/`false` | Copies local DNS records                                    |
| `SYNC_V5_LOCAL_CNAME_RECORDS` | No       | `true`  | `true`/`false` | Copies local CNAME records                                  |
| `SYNC_V5_FLUSH_TABLES`        | No       | `true`  | `true`/`false` | Clears existing data on the secondary (copy target) Pi-hole |

### V6

Sync options for Pi-hole v6.x.

| Environment Variable                 | Required | Default | Example        | Description                                                                                                                                                                                                                                                                                                   |
| ------------------------------------ | -------- | ------- | -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `SYNC_V6_CONFIG`                     | No       | `false` | `true`/`false` | Copies the TOML config file                                                                                                                                                                                                                                                                                   |
| `SYNC_V6_DHCP_LEASES`                | No       | `true`  | `true`/`false` | Copies the DHCP leases                                                                                                                                                                                                                                                                                        |
| `SYNC_V6_GROUP`                      | No       | `true`  | `true`/`false` | Copies groups                                                                                                                                                                                                                                                                                                 |
| `SYNC_V6_ADLIST`                     | No       | `true`  | `true`/`false` | Copies adlists                                                                                                                                                                                                                                                                                                |
| `SYNC_V6_ADLIST_BY_GROUP`            | No       | `true`  | `true`/`false` | Copies adlists by group                                                                                                                                                                                                                                                                                       |
| `SYNC_V6_DOMAINLIST`                 | No       | `true`  | `true`/`false` | Copies domain list                                                                                                                                                                                                                                                                                            |
| `SYNC_V6_DOMAINLIST_BY_GROUP`        | No       | `true`  | `true`/`false` | Copies domain list by group                                                                                                                                                                                                                                                                                   |
| `SYNC_V6_CLIENT`                     | No       | `true`  | `true`/`false` | Copies clients                                                                                                                                                                                                                                                                                                |
| `SYNC_V6_CLIENT_BY_GROUP`            | No       | `true`  | `true`/`false` | Copies clients by group                                                                                                                                                                                                                                                                                       |
| `SYNC_V6_GRAVITY_UPDATE_RETRY_COUNT` | No       | `5`     | `3`            | The number of times to retry updating gravity if it fails. Only used if `UPDATE_GRAVITY` is not set to `false`. Defaults to 5. Uses an exponential backoff; the second attempt will wait a second, the third attempt 2, the fourth attempt 4, the fifth attempt 8, and so on - up to a maximum of 60 seconds. |

## Notify

When to send notifications and how to send them.

| Environment Variable | Required | Default | Example        | Description                                           |
| -------------------- | -------- | ------- | -------------- | ----------------------------------------------------- |
| `NOTIFY_ON_SUCCESS`  | No       | `false` | `true`/`false` | Send a notification if a sync completes successfully. |
| `NOTIFY_ON_FAILURE`  | No       | `true`  | `true`/`false` | Send a notification if a sync fails for any reason.   |

### Smtp

Send notifications via email using SMTP

| Environment Variable   | Required | Default | Example                   | Description                                                                |
| ---------------------- | -------- | ------- | ------------------------- | -------------------------------------------------------------------------- |
| `NOTIFY_SMTP_ENABLED`  | No       | `false` | `true`/`false`            | Send notifications via email.                                              |
| `NOTIFY_SMTP_FROM`     | No       | N/A     | `orbitalsync@example.com` | The email address to send notifications from.                              |
| `NOTIFY_SMTP_TO`       | No       | N/A     | `you@example.com`         | The email address to send notifications to. Can be a comma-separated list. |
| `NOTIFY_SMTP_HOST`     | No       | N/A     | `smtp.example.com`        | The SMTP server host.                                                      |
| `NOTIFY_SMTP_PORT`     | No       | N/A     | `25`/`587`/`465`          | The SMTP server port.                                                      |
| `NOTIFY_SMTP_TLS`      | No       | `false` | `true`/`false`            | Should usually be set to true if using port 465. Otherwise, leave as is.   |
| `NOTIFY_SMTP_USER`     | No       | N/A     | `orbitalsync@example.com` | The SMTP account username.                                                 |
| `NOTIFY_SMTP_PASSWORD` | No       | N/A     | `yourpasswordhere`        | The SMTP account password.                                                 |

### Exceptions

Log exceptions to [Honeybadger](https://www.honeybadger.io) or [Sentry](http://sentry.io/). Used mostly for development or debugging.

| Environment Variable                    | Required | Default | Example                             | Description                                                                                        |
| --------------------------------------- | -------- | ------- | ----------------------------------- | -------------------------------------------------------------------------------------------------- |
| `NOTIFY_EXCEPTIONS_HONEYBADGER_API_KEY` | No       | N/A     | `hbp_xxxxxxxxxxxxxxxxxx`            | Set to use Honeybadger for proper exception recording; mostly useful for development or debugging. |
| `NOTIFY_EXCEPTIONS_SENTRY_DSN`          | No       | N/A     | `https://key@o0.ingest.sentry.io/0` | Set to use Sentry for proper exception recording; mostly useful for development or debugging.      |

<!-- END CONFIG DOCS -->

# Additional Notes

## Docker Secrets

All above configuration options can be set with Docker secrets. In other words, by appending any environment variable with the `_FILE` suffix, you can provide the path to a file that contains the value. For example, `PRIMARY_HOST_PASSWORD_FILE=/run/secrets/pihole_password` would read the primary host password from the file `/run/secrets/pihole_password`. In practice, a `docker-compose.yml` for this configuration might look like:

```yaml
services:
  orbital-sync:
    image: mattwebbio/orbital-sync:latest
    secrets:
      - pihole1_password
      - pihole2_password
    environment:
      - PRIMARY_HOST_BASE_URL=https://pihole1.mydomain.com
      - PRIMARY_HOST_PASSWORD_FILE=/run/secrets/pihole1_password
      - SECONDARY_HOSTS_1_BASE_URL=https://pihole2.mydomain.com
      - SECONDARY_HOSTS_1_PASSWORD_FILE=/run/secrets/pihole2_password
secrets:
  pihole1_password:
    external: true
  pihole2_password:
    external: true
```

If both the `_FILE` and non-`_FILE` versions of an environment variable are set, the `_FILE` version will take precedence.


CHANGELOG.md
### Changelog


## [1.9.0](https://github.com/mattwebbio/orbital-sync/compare/v1.8.4...v1.9.0) (2025-03-24)


### Features

* add support for PiHole v6 ([f07c3c4](https://github.com/mattwebbio/orbital-sync/commit/f07c3c411b7a0d5132198c7b4d1a131ffa1fa005))


### Bug Fixes

* add `SYNC_V6_CONFIG` option and default it to `false` ([b876256](https://github.com/mattwebbio/orbital-sync/commit/b8762561bb87950a6ccb358ce40acc07b54af4cd))
* add retries and exponential backoff for v6 gravity update ([947feb2](https://github.com/mattwebbio/orbital-sync/commit/947feb2955178311e72667177ae5f967a07c8a27))
* bump @babel/runtime from 7.23.9 to 7.26.10 ([f0327ab](https://github.com/mattwebbio/orbital-sync/commit/f0327ab831c7878d59208f3f0fae02deb1229357))
* bump nodemailer and @types/nodemailer ([095ed38](https://github.com/mattwebbio/orbital-sync/commit/095ed38b33762bc8091db9cb7dca8477c11879a5))
* bump undici from 5.28.4 to 5.28.5 ([90d3c1b](https://github.com/mattwebbio/orbital-sync/commit/90d3c1bdce9d05b23478bc547da205527789b5d4))
* retry gravity update if 502, 503, or 504 status is returned ([b2fe070](https://github.com/mattwebbio/orbital-sync/commit/b2fe07074873437f143b8def9acc0c9d90f05040))

## [1.8.4](https://github.com/mattwebbio/orbital-sync/compare/v1.8.3...v1.8.4) (2025-01-11)


### Bug Fixes

* bump @honeybadger-io/js from 6.10.0 to 6.10.1 ([5724009](https://github.com/mattwebbio/orbital-sync/commit/5724009853d88c7783eff9ab0d4f1197a994db78))
* bump fetch-cookie from 3.0.1 to 3.1.0 ([5df230b](https://github.com/mattwebbio/orbital-sync/commit/5df230bebd268a755bc1c8e85b21244bf7cd6097))
* bump node-html-parser from 6.1.13 to 7.0.1 ([4516607](https://github.com/mattwebbio/orbital-sync/commit/451660723238a132b44de259bf832ac263979f2c))

## [1.8.3](https://github.com/mattwebbio/orbital-sync/compare/v1.8.2...v1.8.3) (2024-12-26)


### Bug Fixes

* bump @eslint/plugin-kit from 0.2.0 to 0.2.4 ([bf92191](https://github.com/mattwebbio/orbital-sync/commit/bf921912c13c2a900cfbd2c17612ff0f8be2765a))
* bump @honeybadger-io/js from 6.9.3 to 6.10.0 ([b3219b6](https://github.com/mattwebbio/orbital-sync/commit/b3219b66fce6af5df03ed929dd8008c9cdbb161d))
* bump cross-spawn from 7.0.3 to 7.0.6 ([ea0bd0e](https://github.com/mattwebbio/orbital-sync/commit/ea0bd0eb5e61bc2baf5b50960780c6e628a29aff))
* bump json-schema-to-ts from 3.1.0 to 3.1.1 ([454660c](https://github.com/mattwebbio/orbital-sync/commit/454660cbdbe681bb06a11b46a52773ed13f5e83e))
* bump micromatch from 4.0.5 to 4.0.8 ([45dee5e](https://github.com/mattwebbio/orbital-sync/commit/45dee5ee690308c58d39a45f7aa682dba884e063))
* bump nodemailer from 6.9.14 to 6.9.15 ([47c531a](https://github.com/mattwebbio/orbital-sync/commit/47c531afed9ab1781527402dee6ac63a636e48e2))

## [1.8.2](https://github.com/mattwebbio/orbital-sync/compare/v1.8.1...v1.8.2) (2024-08-05)


### Bug Fixes

* bump @honeybadger-io/js from 6.8.3 to 6.9.1 ([0565088](https://github.com/mattwebbio/orbital-sync/commit/056508818cf312e3b4a0cef1912f4671be954a71))
* bump @honeybadger-io/js from 6.9.1 to 6.9.2 ([a070ba1](https://github.com/mattwebbio/orbital-sync/commit/a070ba13a5cef05ee55a219549d73302d3a4aea3))
* bump @honeybadger-io/js from 6.9.2 to 6.9.3 ([3be6d95](https://github.com/mattwebbio/orbital-sync/commit/3be6d959c9e2fa2260f5861f1a780c7dc2e74ed7))
* bump @sentry/node from 7.108.0 to 7.109.0 ([b8263a9](https://github.com/mattwebbio/orbital-sync/commit/b8263a9e7b1dbac60e15c44b5e9c650719dac7b5))
* bump @sentry/node from 7.109.0 to 7.111.0 ([ace2dbb](https://github.com/mattwebbio/orbital-sync/commit/ace2dbb286939853cdd6ca99fccd798a1171fa9e))
* bump @sentry/node from 7.111.0 to 7.112.2 ([6642c76](https://github.com/mattwebbio/orbital-sync/commit/6642c76295126cbc25e00a7acc5e68285b235f52))
* bump @sentry/node from 7.112.2 to 7.113.0 ([25cecbf](https://github.com/mattwebbio/orbital-sync/commit/25cecbf584a428926efc2cd29a3b871ce80e8804))
* bump @sentry/node from 7.113.0 to 7.114.0 ([c1d3692](https://github.com/mattwebbio/orbital-sync/commit/c1d3692d14680e20b769e21486e7b183a03ca746))
* bump braces from 3.0.2 to 3.0.3 ([cacc15f](https://github.com/mattwebbio/orbital-sync/commit/cacc15fd617073ea3ddbaa86f31407f1861f0e3d))
* bump json-schema-to-ts from 3.0.1 to 3.1.0 ([f76d861](https://github.com/mattwebbio/orbital-sync/commit/f76d861232da4bda67fb25702225133711d2d98d))
* bump node-html-parser from 6.1.12 to 6.1.13 ([aa4c6ae](https://github.com/mattwebbio/orbital-sync/commit/aa4c6aed19cb65ff09adc229342eea5a467b1590))
* bump nodemailer from 6.9.13 to 6.9.14 ([17ea56a](https://github.com/mattwebbio/orbital-sync/commit/17ea56ab6fb8d08b1576166d027508b6e60b129f))
* error "Not authorized" when uploading archives ([c9eea45](https://github.com/mattwebbio/orbital-sync/commit/c9eea45e6b2a806443375d35a5acad3fa53785d3))

## [1.8.1](https://github.com/mattwebbio/orbital-sync/compare/v1.8.0...v1.8.1) (2024-03-25)


### Bug Fixes

* include all `dist` output in NPM release ([fbc9b05](https://github.com/mattwebbio/orbital-sync/commit/fbc9b059da03b33f02cbe84007958bad3efbd930))

## [1.8.0](https://github.com/mattwebbio/orbital-sync/compare/v1.7.0...v1.8.0) (2024-03-25)


### Features

* add distroless image ([a93c38e](https://github.com/mattwebbio/orbital-sync/commit/a93c38e9b0f71033e5ad6bedf00a4bb67c3c2478))


### Bug Fixes

* bump @sentry/node from 7.107.0 to 7.108.0 ([97eeec6](https://github.com/mattwebbio/orbital-sync/commit/97eeec6bcbbce4a235186518b9b38e7b60d5efbc))
* bump json-schema-to-ts from 3.0.0 to 3.0.1 ([9f87443](https://github.com/mattwebbio/orbital-sync/commit/9f874439dd329a7914cda7f5616386de131ff58e))
* bump nodemailer from 6.9.12 to 6.9.13 ([897e9b6](https://github.com/mattwebbio/orbital-sync/commit/897e9b6f53c535aea8c12c8dd8db01776b6b342c))

## [1.7.0](https://github.com/mattwebbio/orbital-sync/compare/v1.6.0...v1.7.0) (2024-03-24)


### Features

* read config values from files ([9febf7b](https://github.com/mattwebbio/orbital-sync/commit/9febf7bf93639fcd259f18d7d094066dca959655))


src/config.ts
import { ErrorNotification } from './notify.js';

export class Config {
  private static _primaryHost?: Host;
  private static _secondaryHosts?: Host[];
  private static _syncOptions?: SyncOptions;
  private static _intervalMinutes?: number;

  static get primaryHost(): Host {
    this._primaryHost ??= new Host(
      this.getRequiredEnv('PRIMARY_HOST_BASE_URL'),
      this.getRequiredEnv('PRIMARY_HOST_PASSWORD'),
      process.env['PRIMARY_HOST_PATH']
    );

    return this._primaryHost;
  }

  static get secondaryHosts(): Host[] {
    if (!this._secondaryHosts) {
      this._secondaryHosts = [
        new Host(
          this.getRequiredEnv('SECONDARY_HOST_1_BASE_URL'),
          this.getRequiredEnv('SECONDARY_HOST_1_PASSWORD'),
          process.env['SECONDARY_HOST_1_PATH']
        )
      ];

      let count = 2;
      while (
        process.env[`SECONDARY_HOST_${count}_BASE_URL`] !== undefined &&
        process.env[`SECONDARY_HOST_${count}_PASSWORD`] !== undefined
      ) {
        this._secondaryHosts.push(
          new Host(
            this.getRequiredEnv(`SECONDARY_HOST_${count}_BASE_URL`),
            this.getRequiredEnv(`SECONDARY_HOST_${count}_PASSWORD`),
            process.env[`SECONDARY_HOST_${count}_PATH`]
          )
        );

        count++;
      }
    }

    return this._secondaryHosts;
  }

  static get allHostUrls(): string[] {
    return [this.primaryHost, ...this.secondaryHosts].map((host) => host.fullUrl);
  }

  static get syncOptions(): SyncOptions {
    this._syncOptions ??= {
      whitelist: process.env['SYNC_WHITELIST'] !== 'false',
      regexWhitelist: process.env['SYNC_REGEX_WHITELIST'] !== 'false',
      blacklist: process.env['SYNC_BLACKLIST'] !== 'false',
      regexlist: process.env['SYNC_REGEXLIST'] !== 'false',
      adlist: process.env['SYNC_ADLIST'] !== 'false',
      client: process.env['SYNC_CLIENT'] !== 'false',
      group: process.env['SYNC_GROUP'] !== 'false',
      auditlog: process.env['SYNC_AUDITLOG'] === 'true',
      staticdhcpleases: process.env['SYNC_STATICDHCPLEASES'] === 'true',
      localdnsrecords: process.env['SYNC_LOCALDNSRECORDS'] !== 'false',
      localcnamerecords: process.env['SYNC_LOCALCNAMERECORDS'] !== 'false',
      flushtables: process.env['SYNC_FLUSHTABLES'] !== 'false'
    };

    return this._syncOptions;
  }

  static get updateGravity(): boolean {
    return process.env['UPDATE_GRAVITY'] !== 'false';
  }

  static get verboseMode(): boolean {
    return process.env['VERBOSE'] === 'true';
  }

  static get notifyOnSuccess(): boolean {
    return process.env['NOTIFY_ON_SUCCESS'] === 'true';
  }

  static get notifyOnFailure(): boolean {
    return process.env['NOTIFY_ON_FAILURE'] !== 'false';
  }

  static get notifyViaSmtp(): boolean {
    return process.env['NOTIFY_VIA_SMTP'] === 'true';
  }

  static get smtpHost(): string {
    return this.getRequiredEnv('SMTP_HOST');
  }

  static get smtpPort(): string {
    return process.env['SMTP_PORT'] ?? '587';
  }

  static get smtpTls(): boolean {
    return process.env['SMTP_TLS'] === 'true';
  }

  static get smtpUser(): string | undefined {
    return process.env['SMTP_USER'];
  }

  static get smtpPassword(): string | undefined {
    return process.env['SMTP_PASSWORD'];
  }

  static get smtpFrom(): string | undefined {
    return process.env['SMTP_FROM'];
  }

  static get smtpTo(): string {
    return this.getRequiredEnv('SMTP_TO');
  }

  static get runOnce(): boolean {
    return process.env['RUN_ONCE'] === 'true';
  }

  static get intervalMinutes(): number {
    if (this._intervalMinutes) return this._intervalMinutes;

    if (process.env['INTERVAL_MINUTES']) {
      const parsed = parseInt(process.env['INTERVAL_MINUTES']);
      if (parsed && parsed > 0) this._intervalMinutes = parsed;
    }

    this._intervalMinutes ??= 30;
    return this._intervalMinutes;
  }

  static get honeybadgerApiKey(): string | undefined {
    return process.env['HONEYBADGER_API_KEY'];
  }

  static get sentryDsn(): string | undefined {
    return process.env['SENTRY_DSN'];
  }

  private static getRequiredEnv(variable: string): string {
    const value = process.env[variable];

    if (value === undefined)
      throw new ErrorNotification({
        message: `The environment variable ${variable} is required but not defined.`,
        exit: true
      });

    return value;
  }
}

export interface SyncOptions {
  whitelist: boolean;
  regexWhitelist: boolean;
  blacklist: boolean;
  regexlist: boolean;
  adlist: boolean;
  client: boolean;
  group: boolean;
  auditlog: boolean;
  staticdhcpleases: boolean;
  localdnsrecords: boolean;
  localcnamerecords: boolean;
  flushtables: boolean;
}

export class Host {
  baseUrl: string;
  path: string;
  fullUrl: string;
  password: string;

  private static pathExtractor = RegExp('^(http[s]?:+//[^/s]+)([/]?[^?#]+)?');

  constructor(baseUrl: string, password: string, path = '/admin') {
    if (path && !path.startsWith('/')) {
      path = '/' + path;
    }

    const includedPath = Host.pathExtractor.exec(baseUrl);

    if (includedPath && includedPath[1] && includedPath[2]) {
      baseUrl = includedPath[1];
      path = (this.trimTrailingSlash(includedPath[2]) ?? '') + path;
    }

    this.baseUrl = baseUrl;
    this.password = password;
    this.path = this.trimTrailingSlash(path);
    this.fullUrl = this.baseUrl + this.path;
  }

  private trimTrailingSlash(url: string): string {
    return url.endsWith('/') ? url.slice(0, url.length - 1) : url;
  }
}
